packages/config-loader/src/lib/transform/include.ts

```diff
@@ -0,0 +1,124 @@
+  '.yaml': async content => yaml.parse(content),
+  '.yml': async content => yaml.parse(content),
+};
+
+/**
+ * Transforms a secret description into the actual secret value.
+ */
+export function createIncludeTransform(
+  env: EnvFunc,
+  readFile: ReadFileFunc,
+): TransformFunc {
+  return async (input: JsonValue, baseDir: string) => {
+    if (!isObject(input)) {
+      return { applied: false };
+    }
+    // Check if there's any key that starts with a '$', in that case we treat
+    // this entire object as a secret.
+    const [secretKey] = Object.keys(input).filter(key => key.startsWith('$'));
+    if (secretKey) {
+      if (Object.keys(input).length !== 1) {
+        throw new Error(
+          `include key ${secretKey} should not have adjacent keys`,
+        );
+      }
+    } else {
+      return { applied: false };
+    }
+
+    const secretValue = input[secretKey];
+    if (typeof secretValue !== 'string') {
+      throw new Error(`${secretKey} include value is not a string`);
+    }
+
+    switch (secretKey) {
+      case '$file':
+        try {
+          const value = await readFile(resolvePath(baseDir, secretValue));
+          return { applied: true, value };
+        } catch (error) {
+          throw new Error(`failed to read file ${secretValue}, ${error}`);
+        }
+      case '$env':
+        try {
+          return { applied: true, value: await env(secretValue) };
+        } catch (error) {
+          throw new Error(`failed to read env ${secretValue}, ${error}`);
+        }
+
+      case '$include': {
+        const [filePath, dataPath] = secretValue.split(/#(.*)/);
+
+        const ext = extname(filePath);
+        const parser = includeFileParser[ext];
+        if (!parser) {
+          throw new Error(
+            `no configuration parser available for extension ${ext}`,
+          );
+        }
+
+        const path = resolvePath(baseDir, filePath);
+        const content = await readFile(path);
+        const newBaseDir = dirname(path);
+
+        const parts = dataPath ? dataPath.split('.') : [];
+
+        let value: JsonValue | undefined;
+        try {
+          value = await parser(content);
+        } catch (error) {
+          throw new Error(
+            `failed to parse included file ${filePath}, ${error}`,
+          );
+        }
+
+        // This bit handles selecting a subtree in the included file, if a path was provided after a #
+        for (const [index, part] of parts.entries()) {
+          if (!isObject(value)) {
+            const errPath = parts.slice(0, index).join('.');
+            throw new Error(
+              `value at '${errPath}' in included file ${filePath} is not an object`,
+            );
+          }
+          value = value[part];
+        }
+
+        return {
+          applied: true,
+          value,
+          newBaseDir: newBaseDir !== baseDir ? newBaseDir : undefined,
+        };
+      }
+
+      default:
+        throw new Error(`unknown secret ${secretKey}`);
```

Comment:
about this file, still referring to them as secrets throughout rather than includes or placeholders